### PR TITLE
Extension of the framework nuclides

### PR DESCRIPTION
**Pull request by jakehader**
## Description

Update to the nuclides that are built into the ARMI framework. This expands the set of nuclide bases to include 4614 nuclides and imports the half-life data from the chart of the nuclides from a combination of the RIPL-3 database and from the IAEA chart of the nuclides.

Other Changes:

* Each `Element` now has a `phase` and a `group`.
* `Element.nuclideBases` becomes `Element.nuclides` - This is an API-breaking change.

---

## Checklist

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

**Comment by jakehader** [comment]
This will take some time to review, but please see the attached ipython notebook that can be used to check the data that is being proposed in this PR against the IAEA Chart of the Nuclides - https://www-nds.iaea.org/relnsd/vcharthtml/VChartHTML.html

Note that GitHub doesn't allow file uploads of `.ipynb` so the extension is changed to `.txt`.

[iaea-check.txt](https://github.com/terrapower/armi/files/10109505/iaea-check.txt)

**Comment by jakehader** [comment]
@john-science - I am not sure why the tests didn't run. Is this something that has been failing? Unit tests pass locally for me.

**Comment by jakehader** [comment]
image
[mass-comparisons.txt](https://github.com/terrapower/armi/files/10118092/mass-comparisons.txt)
[nusf-comparisons.txt](https://github.com/terrapower/armi/files/10118093/nusf-comparisons.txt)
[abundance-comparisons.txt](https://github.com/terrapower/armi/files/10118094/abundance-comparisons.txt)
[halflife-comparisons.txt](https://github.com/terrapower/armi/files/10118095/halflife-comparisons.txt)

**Comment by john-science** [comment]
@jakehader Assuming we get the unit tests working and all that...

> Do you think this will require any breaking to our downstream projects? 

**NOTE**: @opotowsky @sombrereau @drewj-usnctech

**Comment by john-science** [comment]
@jakehader @ntouran How much do we expect this to slow down ARMI?  This is a _lot_ more nuclides.

**Comment by opotowsky** [comment]
> @john-science - I am not sure why the tests didn't run. Is this something that has been failing? Unit tests pass locally for me.

```
  File "/home/runner/work/armi/armi/armi/physics/neutronics/fissionProductModel/lumpedFissionProduct.py", line 506, in createLFPsFromFile
    lfp = self._readOneLFP(lfpLines)
  File "/home/runner/work/armi/armi/armi/physics/neutronics/fissionProductModel/lumpedFissionProduct.py", line 576, in _readOneLFP
    nuc = nuclideBases.byMccId[nucLibId]
KeyError: 'GE73 5'
```

Once this is handled I can run some tests downstream. They also fail on the above unit test failure

**Comment by drewj-usnctech** [comment]
Thanks for the ping. No problems on our end

**Comment by jakehader** [comment]
> Thanks for the ping. No problems on our end

Any recommendations or other needs coming from this? I'm not sure if you're still using Serpent @drewj-usnctech, but maybe a run down on if the nuclide bases work for getting Serpent IDs correctly would be nice. I am not sure if it's the same between ENDF/B-VII.0, ENDF/B-VII.1, and ENDF/B-VIII.0. Is there any good resource for this?

**Comment by drewj-usnctech** [comment]
> > Thanks for the ping. No problems on our end
> 
> Any recommendations or other needs coming from this? I'm not sure if you're still using Serpent @drewj-usnctech, but maybe a run down on if the nuclide bases work for getting Serpent IDs correctly would be nice. I am not sure if it's the same between ENDF/B-VII.0, ENDF/B-VII.1, and ENDF/B-VIII.0. Is there any good resource for this?

Except for some rare exceptions, Serpent and MCNP ids align very well. We haven't hit major stumbling blocks

**Comment by jakehader** [comment]
> > So the way I read this is that you do not use nuclideBase.getSerpentId()? Would you care if we removed this?
> 
> Sorry that was not the intent I meant to express. We do use `nuclideBase.getSerpentId` to get the more human readable nuclide names (`Al-27`) vs. the more traditional `ZZAAA` form (`13026`)
> 
> To the point of
> 
> > I am not sure if it's the same between ENDF/B-VII.0, ENDF/B-VII.1, and ENDF/B-VIII.0
> 
> I'm pretty sure the only thing that changes from evaluation to evaluation is the library suffix, not so much the nuclide naming structure. So you might see `13026.03c` in one eval, but `13026.08c` later, both at the same temperature. I might be missing something in my interpretation of your comment though

Okay thank you for that. Don't want to mess things up for you so I will definitely keep this!

File: armi/nucDirectory/elements.py [hunk 1]
```diff
@@ -25,124 +27,155 @@
...
-        for nuc in self.nuclideBases:
+        for nuc in sorted(self.nuclides):
             yield nuc
 
     def append(self, nuclide):
-        self.nuclideBases.append(nuclide)
+        """Assigns and sorts the nuclide to the element and ensures no duplicates."""
+        from armi.nucDirectory.nuclideBases import NuclideBase
+
+        if nuclide in self.nuclides or not isinstance(nuclide, NuclideBase):
```
Comment (keckler) [inline armi/nucDirectory/elements.py:105]: I understand just returning if the nuclide is already in the list, but do you think we should instead fail if the `nuclide` isn't a valid instance?

File: armi/nucDirectory/elements.py [hunk 2]
```diff
@@ -25,124 +27,155 @@
...
+        return any([nuc.abundance > 0.0 for nuc in self.nuclides])
 
     def getNaturalIsotopics(self):
+        """Return a list of nuclides that are naturally occurring for this element."""
+        return [nuc for nuc in self.nuclides if nuc.abundance > 0.0 and nuc.a > 0]
+
+    def isHeavyMetal(self):
         """
-        Return the nuclide bases of any naturally-occurring isotopes of this element.
+        Return True if the atomic number of greater than 89 (i.e., Z > 89).
```
Comment (keckler) [inline armi/nucDirectory/elements.py:120]: Perhaps you don't want to specifically cite "89" in the docstring, to avoid it becoming out of sync with `HEAVY_METAL_CUTOFF_Z`.

File: armi/resources/mc2Nuclides.yaml [hunk 3]
```diff
@@ -1,1724 +0,0 @@
-AC225:
```
Comment (keckler) [inline armi/resources/mc2Nuclides.yaml:1]: There is a stray reference to this file in the armi/reactor/blueprints/isotoopicOptions module.

image

File: armi/resources/burn-chain.yaml [hunk 4]
```diff
@@ -1,5 +1,4 @@
 AM241:
-- nuSF: 2.5
```
Comment (keckler) [inline armi/resources/burn-chain.yaml:2]: Can you comment on why all of the nuSF values have been stripped out of this file?

File: armi/resources/nuclides.dat [hunk 5]
```diff
@@ -0,0 +1,4664 @@
...
+# This file contains a set of 4614 nuclides and their associated metadata
+# of atomic number (Z), number of neutrons (N), mass number (A), stability
+# state number (S), Mass (amu), natural abundance (%), half-life (sec), and
+# the spontaneous fission neutron yield (nu-bar).
+# 
+# The stability state is either a zero (0), representing the ground state, or
+# higher (e.g., 1, 2, 3, etc.) representing excited/metastable states. In
+# most instances, the ground state represents the most commonly produced isotope
+# from fission or activation, but there are some known exceptions. For example,
+# AM242M is a more common nuclide produced from activation than the group state
```
Comment (keckler) [inline armi/resources/nuclides.dat:13]: Should say "ground state", I believe

File: armi/resources/nuclides.dat [hunk 6]
```diff
@@ -0,0 +1,4664 @@
...
+# The nuclide natural abundance data is acquired from (1) by parsing the "abundance"
+# data file. The original source of this data is from the Brookhaven Nuclear Wallet Cards (2).       
+#
+# The half-life data for each nuclide is provided as a direct output from reading and 
+# processing the RIPL-3 "levels" data. A half-life provided as `inf` represents a stable 
+# nuclide and internally this is processed as an "infinity" using the Python numpy package.
+# Note that for some nuclides where the halflife is not known (i.e., TI48M), the halflife
+# is set to a fixed value of 11.111 seconds. This value is set arbitrarily, but aligns with
+# an assumption made in TerraPower's internal depletion solver. In almost all instances this
+# this assumption will not be consequential for practical reactor physics applications, but this
```
Comment (keckler) [inline armi/resources/nuclides.dat:43]: Duplicate "this" should be removed.

File: armi/resources/burn-chain.yaml (hunk 4 above)
Comment (keckler) [reply armi/resources/burn-chain.yaml:2]: I guess it is to avoid duplication with the data now stored in the nuclides.dat file?

**Review by keckler — COMMENTED** [review]
This is only a partial review. Just hitting the low-hanging fruit. 

I have NOT yet reviewed nuclideBases.py

**Comment by keckler** [comment]
I think this docstring needs to be modified:
image

**Comment by jakehader** [comment]
Thanks @keckler! Looking at the comments now to address

File: armi/resources/endf_mf_nums.txt [hunk 7]
```diff
@@ -1,25 +0,0 @@
-MF #    Quantity    Quantity Description (http://www.nndc.bnl.gov/exfor/help7.jsp)
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/resources/endf_mf_nums.txt:1]: See comment on other file below

File: armi/resources/burn-chain.yaml (hunk 4 above)
Comment (jakehader) [reply armi/resources/burn-chain.yaml:2]: Yes basically that data is in nuclides.dat now and this can be treated as a way to override data that is loaded in with this file. This goes for halflife as well.

**Comment by keckler** [comment]
I can also confirm that all of the half-life data for ground state isotopes compares well against the IAEA data that was referenced above. In this case, "well" means that none of the half-lives are more than 1e-8% different than each other, with the exception of N11, which has 2 half-lives produced by the IAEA's API. One of those values agrees very well, and the other is off by >200%. However, that isotope has a half-life down near 1e-22 seconds.

Still working on verifying the following:
- [ ] Half-lives of excited states
- [ ] Mass
- [ ] Abundance
- [ ] Spontaneous fission yields

I will get to these tomorrow.

File: armi/resources/elements.dat [hunk 8]
```diff
@@ -1,118 +1,132 @@
...
+4	Be	Beryllium                   SOLID    ALKALINE_EARTH_METAL
+5	B	Boron                       SOLID    METALLOID
+6	C	Carbon                      SOLID    NONMETAL
+7	N	Nitrogen                    GAS      NONMETAL
+8	O	Oxygen                      GAS      NONMETAL
+9	F	Fluorine                    GAS      HALOGEN
+10	Ne	Neon                        GAS      NOBEL_GAS
+11	Na	Sodium                      SOLID    ALKALI_METAL
+12	Mg	Magnesium                   SOLID    ALKALINE_EARTH_METAL
+13	Al	Aluminium                   SOLID    POST_TRANSITION_METAL
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/resources/elements.dat:25]: Nice - I am updating...

File: armi/resources/nuclides.dat [hunk 9]
```diff
@@ -0,0 +1,4664 @@
...
+# the half-life is set to a fixed value of 11.111 seconds. This value is set arbitrarily, 
+# but aligns with an assumption made in TerraPower's internal depletion solver. In almost 
+# all instances this assumption will not be consequential for practical reactor physics 
+# applications, but this should be verified on a case-by-case basis. 
+#
+# Sources: (1) Reference Input Parameter Library (RIPL-3) - 
+#              https://www-nds.iaea.org/RIPL-3/
+#          (2) J.K. Tuli, Nuclear Wallet Cards (NNDC, Brookhaven National Laboratory), 2005.
+# ---------------------------------------------------------------------
+Z	N	A	S	El	  Mass, amu	        Abundance, %      Half-life (sec)   Spontaneous Fission Yield (nu-bar)
```
Comment (keckler) [inline armi/resources/nuclides.dat:50]: The abundances in this file appear to be fractions, not percent, as is currently indicated in the header line.

File: armi/resources/nuclides.dat [hunk 10]
```diff
@@ -0,0 +1,4664 @@
...
+# data file. The original source of this data is from the Brookhaven Nuclear Wallet Cards (2).       
+#
+# The half-life data for each nuclide is provided as a direct output from reading and 
+# processing the RIPL-3 "levels" data. A half-life provided as `inf` represents a stable 
+# nuclide and internally this is processed as "infinity" using the Python numpy package. 
+# Note that for some nuclides where the half-life is not known (i.e., TI48M), 
+# the half-life is set to a fixed value of 11.111 seconds. This value is set arbitrarily, 
+# but aligns with an assumption made in TerraPower's internal depletion solver. In almost 
+# all instances this assumption will not be consequential for practical reactor physics 
+# applications, but this should be verified on a case-by-case basis.
```
Comment (keckler) [inline armi/resources/nuclides.dat:44]: Please add some text describing the source of the nu-bar data, in the same way that you have done for the other items.

File: armi/resources/elements.dat [hunk 11]
```diff
@@ -1,118 +1,132 @@
...
-114	Uuq	Ununquadium
-115	Uup	Ununpentium
-116	Uuh	Ununhexium
-117	Uus	Ununseptium
-118	Uuo	Ununoctium
+# ---------------------------------------------------------------------
+# ELEMENTAL DATA
+# ---------------------------------------------------------------------
+# This file contains the complete set of elements from the
+# Periodic Table, with chemical phase and group metadata. The
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/resources/elements.dat]: Updated

File: armi/resources/mc2Nuclides.yaml (hunk 3 above)
Comment (jakehader) [reply armi/resources/mc2Nuclides.yaml:1]: Fixed!

File: armi/resources/nuclides.dat (hunk 6 above)
Comment (jakehader) [reply armi/resources/nuclides.dat]: Fixed

File: armi/resources/nuclides.dat (hunk 10 above)
Comment (jakehader) [reply armi/resources/nuclides.dat:44]: Will do thank you

File: armi/resources/nuclides.dat (hunk 9 above)
Comment (jakehader) [reply armi/resources/nuclides.dat:50]: Thanks! Originally I had percents but changes to fractions so I'll update.

**Comment by keckler** [comment]
The current status of the review on the `nuclides.dat` file is as follows:
1. The half-life data that is currently in `nuclides.dat` does not appear to come from the RIPL-3 database, as is claimed in the header of `nuclides.dat`. It appears to align with the half-life data from the IAEA chart of the nuclides. Either the data in `nuclides.dat` should be updated to align with the RIPL-3 database as is claimed, or the text at the top of the `nuclides.dat` file should be updated to reflect where the data actually comes from.
2. The spontaneous fission yield data appears to align with its underlying source, however its source is not documented. The `nuclides.dat` header should be updated to reflect the source of this data (such a comment was suggested previously as well). This data cannot be compared against the IAEA chart of the nuclides because such data is not available from that source.
3. The mass and abundance data appear to align well with both their underlying sources and with the equivalent data from the IAEA chart of the nuclides.

My final request on this topic is the following:
> The script used to process the RIPL-3 data and produce the `nuclides.dat` file should be uploaded to the repo. A user should be able to specify a local directory containing their own RIPL files, and should be able to reproduce the `nuclides.dat` file. This will help to retain the traceability, verifiability, and reproducibility of the data that is in the `nuclides.dat` file

Therefore, once the following items are successfully completed, I will be happy with `nuclides.dat`:
- [ ] Update the half-life data and/or source
- [ ] List the source of the spontaneous fission yield data
- [ ] Upload the script that can be used to reproduce the `nuclides.dat`

**Comment by jakehader** [comment]
@keckler  - ready for your final checks, sorry for all the commits it took some time to decouple the fission product model code. I opened https://github.com/terrapower/armi/issues/1013 in response to your last comment.

**Comment by john-science** [comment]
Looks like we lost 40 lines of code coverage somewhere.

File: armi/physics/neutronics/fissionProductModel/lumpedFissionProduct.py [hunk 12]
```diff
@@ -250,7 +250,9 @@ def getLanthanideFraction(self):
 
         # sum up all of the nuclides that are XE or KR
```
[the remark this answers is not in the document]
Comment (keckler) [reply armi/physics/neutronics/fissionProductModel/lumpedFissionProduct.py:250]: @jakehader Can you please address this comment?

File: armi/physics/neutronics/globalFlux/globalFluxInterface.py [hunk 13]
```diff
@@ -352,7 +352,7 @@ def fromReactor(self, reactor: reactors.Reactor):
         self.symmetry = reactor.core.symmetry
 
         cycleNodeStamp = f"{reactor.p.cycle:03d}{reactor.p.timeNode:03d}"
-        if self.savePhysicsFilesList is not None:
+        if self.savePhysicsFilesList:
```
Comment (keckler) [inline armi/physics/neutronics/globalFlux/globalFluxInterface.py:355]: lol, what a funny bug. Someone was trying to _avoid_ short-circuiting boolean evaluation but they ended up allowing `False` values to be seen as `True`.

Nice catch!

File: armi/settings/fwSettings/globalSettings.py [hunk 14]
```diff
@@ -129,6 +130,18 @@ def defineSettings() -> List[setting.Setting]:
...
+            default=True,
+            label="Initialize Burn Chain",
+            description=(
+                f"This setting is paired with the `{CONF_BURN_CHAIN_FILE_NAME}` setting. "
+                f"When enabled, this will initialize the burn-chain on initializing the case and "
+                f"is required for running depletion calculations where the transmutations and decays "
+                f"are controlled by the framework. If an external software, such as ORIGEN, contains "
+                f"data for the burn-chain already embedded then this may be disabled."
+            ),
+        ),
```
Comment (keckler) [inline armi/settings/fwSettings/globalSettings.py:144]: Not necessary, but maybe you want to add a schema to only allow booleans?

**Review by keckler — COMMENTED** [review]
I haven't quite finished this review, but this is what I have for now.

I will work on this again first thing tomorrow.

File: armi/physics/neutronics/globalFlux/globalFluxInterface.py (hunk 13 above)
Comment (jakehader) [reply armi/physics/neutronics/globalFlux/globalFluxInterface.py:355]: Yeah this was causing a bug down stream in our plugins as I was running some tests.

File: armi/physics/neutronics/fissionProductModel/lumpedFissionProduct.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/physics/neutronics/fissionProductModel/lumpedFissionProduct.py:250]: Removing the outdated comment now.

File: armi/resources/referenceFissionProducts.dat [hunk 15]
```diff
@@ -145,43 +134,30 @@
...
 13          LFP35 GD1587 6.0000E-05
 13          LFP35 GD1607 1.0000E-05
 13          LFP35 TB1597 3.0000E-05
-13          LFP35 TB1607 0.0000E+00
-13          LFP35 TB1617 0.0000E+00
-13          LFP35 DY1607 0.0000E+00
-13          LFP35 DY1617 0.0000E+00
-13          LFP35 DY1627 0.0000E+00
-13          LFP35 DY1637 0.0000E+00
-13          LFP35 DY1647 0.0000E+00
```
Comment (jakehader) [inline armi/resources/referenceFissionProducts.dat:161]: Removed these because they were zero and had no impact and removed Y90 because its not a nuclide that is modeled in MC2-3.

File: armi/materials/material.py [hunk 16]
```diff
@@ -310,8 +310,16 @@ def adjustMassFrac(self, nuclideName: str, massFraction: float) -> None:
         updatedMassDensities = molesPerCC * atomicMasses
         updatedDensity = updatedMassDensities.sum()
         massFracs = updatedMassDensities / updatedDensity
-        self.p.massFrac = {nuc: weight for nuc, weight in zip(nucsNames, massFracs)}
 
+        # There can be small floating point issues with division, so this
+        # rounds the mass fractions to the nearest 10 decimals.
+        massFracs = massFracs.round(decimals=10)
+        if massFracs.sum() != 1.0:
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/materials/material.py:317]: Implementing now.

File: armi/nucDirectory/__init__.py [hunk 17]
```diff
@@ -99,8 +98,8 @@
 number (A), the natural abundance, and all of the decay and transmutation modes (well, ARMI's
 decay and transmutation modes).
 
-Nuclide names, labels and MC\*\*2 IDs
--------------------------------------
+Nuclide names, labels, and IDs
+------------------------
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/__init__.py:102]: Updating now.

File: armi/settings/fwSettings/globalSettings.py (hunk 14 above)
Comment (jakehader) [reply armi/settings/fwSettings/globalSettings.py:144]: Implemented.

File: armi/nucDirectory/transmutations.py [hunk 18]
```diff
@@ -136,7 +137,12 @@ def __init__(self, parent, dataDict):
         self.productParticle = dataDict.get(
             "productParticle", PRODUCT_PARTICLES.get(self.type)
         )
-        self.branch = dataDict.get("branch", 1.0)
+        self.branch = dataDict.get("branch", None)
+        if self.branch is None:
+            self.branch = 1.0
+            runLog.info(
+                f"The branching ratio for {self} was not defined and is assumed to be 1.0."
+            )
```
Comment (keckler) [inline armi/nucDirectory/transmutations.py:145]: Very nice change! Much more transparent!

File: armi/nucDirectory/transmutations.py [hunk 19]
```diff
@@ -212,7 +218,21 @@ class DecayMode(Transmutable):
...
+        userHalfLife = dataDict.get("halfLifeInSeconds", None)
+        if userHalfLife:
+            if userHalfLife != parent.halflife:
+                runLog.info(
+                    f"Half-life provided for {self} will be updated from "
+                    f"{parent.halflife:<15.11e} to {userHalfLife:<15.11e} seconds based on "
+                    f"user provided burn-chain data."
+                )
+
+                self.halfLifeInSeconds = dataDict["halfLifeInSeconds"]
```
Comment (keckler) [inline armi/nucDirectory/transmutations.py:235]: You don't have to do the dict lookup again here. You already have this value as `userHalfLife`.

File: armi/nucDirectory/nuclideBases.py [hunk 20]
```diff
@@ -825,7 +445,7 @@ def getDecay(self, decayType):
         return None
 
     def isFissile(self):
-        r"""Determine if the nuclide is fissile.
+        """Determine if the nuclide is fissile.
 
         Determines if the nuclide is fissle.
```
Comment (keckler) [inline armi/nucDirectory/nuclideBases.py:450]: This part of the docstring is duplicated.

File: armi/nucDirectory/nuclideBases.py [hunk 21]
```diff
@@ -143,625 +141,233 @@ def getAttributes(nuc):
...
 
+class NuclideInterface:
+    """An abstract nuclide implementation which defining various methods required for a nuclide object."""
 
-def isotopes(z):
-    return elements.byZ[z].nuclideBases
-
+    def getDatabaseName(self):
+        """Return the the nuclide label for the ARMI database (i.e. "nPu239")"""
+        raise NotImplementedError
```
Comment (keckler) [inline armi/nucDirectory/nuclideBases.py:150]: As a general comment, a lot of the machinery in this module could potentially be cleaned up using the `abc` package: https://docs.python.org/3/library/abc.html

**Review by keckler — COMMENTED** [review]
Another batch of comments for you!

File: armi/nucDirectory/nuclideBases.py (hunk 20 above)
Comment (jakehader) [reply armi/nucDirectory/nuclideBases.py:450]: Fixed.

File: armi/nucDirectory/tests/test_elements.py [hunk 22]
```diff
@@ -66,8 +63,22 @@ def test_element_addedElementAppearsInElementList(self):
...
         with mockRunLogs.BufferLog():
-            elements.destroy()
+            nuclideBases.destroyGlobalNuclides()
             elements.factory()
+            nuclideBases.factory()
+            # Ensure that the burn chain data is initialized after clearing
+            # out the nuclide data and reinitializing it.
+            nuclideBases.burnChainImposed = False
+            with open(os.path.join(RES, "burn-chain.yaml"), "r") as burnChainStream:
+                nuclideBases.imposeBurnChain(burnChainStream)
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/tests/test_elements.py:73]: Yeah... since this test updates the global elements by adding "bacon" it will mess with later tests that check for the number of defined elements if the globals are not reset back to the original state. It's a quirk of nuclide and element data being global after ARMI is imported.

File: armi/nucDirectory/elements.py [hunk 23]
```diff
@@ -280,23 +358,44 @@ def factory():
...
+        raise ValueError(f"{element} has already been added and cannot be duplicated.")
+
+    byZ[element.z] = element
+    byName[element.name] = element
+    bySymbol[element.symbol] = element
+
+
+def destroyGlobalElements():
+    """Delete all global elements."""
+    global byZ
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/elements.py:392]: I tested this out and clearing the dictionaries in the module were not done unless I put these `global` statements here. I think this tells Python that the variables in this function are actually globals on the module-level versus local instances. Maybe this example is relevant - https://www.programiz.com/python-programming/global-keyword

File: armi/nucDirectory/transmutations.py (hunk 19 above)
Comment (jakehader) [reply armi/nucDirectory/transmutations.py:235]: Fixed.

File: armi/resources/mcc-nuclides.yaml [hunk 24]
```diff
@@ -0,0 +1,2468 @@
+AC225:
+  ENDF/B-V.2: null
+  ENDF/B-VII.0: AC2257
+  a: 225
+  state: 0
+  z: 89
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/resources/mcc-nuclides.yaml:6]: You're right. With the changes in `__readMCCNuclideData` in `nuclideBases` these are entirely unused now. Deleting.

File: armi/nucDirectory/nuclideBases.py (hunk 21 above)
Comment (jakehader) [reply armi/nucDirectory/nuclideBases.py:150]: Good comment - I will consider this in the future.

File: armi/nucDirectory/elements.py [hunk 25]
```diff
@@ -13,136 +13,257 @@
...
-        totalAbundance = 0.0
-        for nuc in self.nuclideBases:
-            totalAbundance += nuc.abundance
-        return totalAbundance > 0.0
+        """Return True if the element is occurs in nature."""
+        return any([nuc.abundance > 0.0 for nuc in self.nuclides])
 
     def getNaturalIsotopics(self):
+        """Return a list of nuclides that are naturally occurring for this element."""
+        return [nuc for nuc in self.nuclides if nuc.abundance > 0.0 and nuc.a > 0]
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/elements.py:204]: `nuclides.dat` does not have any `a==0` nuclides so you are correct there. Within `nuclideBases` we add natural "nuclides", lumped fission products, and dummy nuclides to the instances. The natural nuclides get added to the elements they belong to and this `nuc.a>0` is to filter those. I will change this to an `instance(nuc, NaturalNuclideBase)` to make this clearer.

[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/elements.py:204]: Actually, I cannot import `nuclideBases` here to do and `ImportError`

```
ImportError while loading conftest 'C:\Users\jhader\codes\nala\framework\armi\conftest.py'.
..\..\__init__.py:61: in <module>
    import armi._bootstrap
..\..\_bootstrap.py:64: in <module>
    nuclideBases.factory()
..\nuclideBases.py:1118: in factory
    __addNaturalNuclideBases()
..\nuclideBases.py:1169: in __addNaturalNuclideBases
    NaturalNuclideBase(element.symbol, element)
..\nuclideBases.py:671: in __init__
    [nn.weight * nn.abundance for nn in element.getNaturalIsotopics()]
..\elements.py:206: in getNaturalIsotopics
    from armi import nuclideBases
E   ImportError: cannot import name 'nuclideBases' from partially initialized module 'armi' (most likely due to a circular import) (C:\Users\jhader\codes\nala\framework\armi\__init__.py)
```

[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/elements.py:204]: Adding a note for this instead in the docstring:

```
    def getNaturalIsotopics(self):
        """
        Return a list of nuclides that are naturally occurring for this element.
        
        Notes
        -----
        This method will filter out any NaturalNuclideBases from the `nuclides`
        attribute.
        """
        return [nuc for nuc in self.nuclides if nuc.abundance > 0.0 and nuc.a > 0]
```

File: armi/nucDirectory/elements.py [hunk 26]
```diff
@@ -13,136 +13,257 @@
 # limitations under the License.
 
 """
-Deals with elements of the periodic table.
+This module provides fundamental element information to be used throughout the framework
+and applications.
+
+The element class structure is outlined :ref:`here <elements-class-diagram>`.
+
+.. _elements-class-diagram:
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/elements.py:21]: Yeah, so the code below builds a class diagram programmatically and this `_elements-class-diagram` is just a link to the pointer defined in the line above. This is the same as in `nuclideBases` so I copied here;

```
The nuclide class structure is outlined :ref:`here <nuclide-bases-class-diagram>`.

.. _nuclide-bases-class-diagram:

.. pyreverse:: armi.nucDirectory.nuclideBases
    :align: center
    :width: 75%

    Class inheritance diagram for :py:class:`INuclide`.
```

**Review by keckler — APPROVED** [review]
Thanks for making those changes, and thanks for all your hard work on this huge effort!

File: armi/reactor/composites.py [hunk 27]
```diff
@@ -1681,10 +1681,6 @@ def getLumpedFissionProductCollection(self):
     def setLumpedFissionProducts(self, lfpCollection):
         self._lumpedFissionProducts = lfpCollection
 
-    def setChildrenLumpedFissionProducts(self, lfpCollection):
-        for c in self.getChildren():
-            c.setLumpedFissionProducts(lfpCollection)
-
```
[the remark this answers is not in the document]
Comment (keckler) [reply armi/reactor/composites.py:1687]: @jakehader 
Can you confirm this? I think you might have accidentally deleted a method that is still in use (see the linked code above)

File: armi/_bootstrap.py [hunk 28]
```diff
@@ -34,9 +34,6 @@
     )
 
 
-from armi import context
```
Comment (john-science) [inline armi/_bootstrap.py:37]: Awesome!

How did this PR allow us to remove `context` from our bootstrapping?  I love the change, but don't really see the connection.

File: armi/reactor/composites.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/composites.py:1687]: Jake? Thoughts?

File: armi/nuclearDataIO/tests/test_ripl.py [hunk 29]
```diff
@@ -1,160 +0,0 @@
...
-        self.assertTrue(os.path.exists(ripDecayFiles[0]))
-
-    def test_RiplDecayFile(self):
-        testDecayConstants = ripl.getNuclideDecayConstants(
-            os.path.join(THIS_DIR, "z036.dat")
-        )
-        kr69 = testDecayConstants[nuclideBases.byName["KR69"]]
-        self.assertAlmostEqual(1.0 - kr69 / 2.53e01, 0, 3)
-
-        for nucName, refDecayConstant in REF_KR_DECAY_CONSTANTS:
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nuclearDataIO/tests/test_ripl.py:133]: Just added these back. Thank you for the suggestions!

File: armi/reactor/composites.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/reactor/composites.py:1687]: Adding back. This will be removed/updated in my follow-on PR.

File: armi/resources/mcc-nuclides.yaml [hunk 30]
```diff
@@ -0,0 +1,2468 @@
+AC225:
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/resources/mcc-nuclides.yaml:1]: Added.

File: armi/nucDirectory/tests/test_nuclideBases.py [hunk 31]
```diff
@@ -256,9 +242,9 @@ def test_nucBases_AllDatabaseNamesAreUnique(self):
     def test_nucBases_Am242m(self):
         am242m = nuclideBases.byName["AM242"]
         self.assertEqual(am242m, nuclideBases.byName["AM242M"])
-        self.assertEqual("nAm242", am242m.getDatabaseName())
+        self.assertEqual("nAm242m", am242m.getDatabaseName())
         self.assertEqual(am242m, nuclideBases.byDBName["nAm242"])
-        self.assertAlmostEqual(am242m.weight, 242.05954949)
+        self.assertAlmostEqual(am242m.weight, 242.059601666)
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/nucDirectory/tests/test_nuclideBases.py:250]: The `mcc-nuclides` is the data that should be consistent with `getMcc2Id` and `getMcc3Id`. Adding tests for this now.

File: armi/_bootstrap.py (hunk 28 above)
Comment (jakehader) [reply armi/_bootstrap.py]: Added back and created an issue on this to track #1037

File: armi/physics/executers.py [hunk 32]
```diff
@@ -191,7 +191,7 @@ def run(self):
         if self.options.savePhysicsFiles:
             outputDir = os.path.join(pathTools.armiAbsPath(os.getcwd()), state, dirName)
         else:
-            outputDir = None
+            outputDir = pathTools.armiAbsPath(os.getcwd())
```
Comment (john-science) [inline armi/physics/executers.py:194]: **NOTE**: This line is added as an unrelated bug fix, to a line added [here](https://github.com/terrapower/armi/commit/d33b7f9978f857da538b37789082b6db48c00e9d#diff-4a5f481998dc37c62ef0f03e848e807414124dc5f9ffc742d9ece1f2ad52c5d4R194).

**Review by john-science — APPROVED** [review]
